### PR TITLE
Log user logins and deactivate password after 3 incorrect logins

### PR DESCRIPTION
**Pull request by punchagan**
This is a work in progress, but I'm creating a PR so that the initial code can be reviewed and modified if required. 

- Successful and unsuccessful logins are logged using the `LoginLog` model
- A user's password is disabled after 3 incorrect attempts for 20 minutes, after which the user can try to login, again. 

- [ ] Test plan | Unit test | Integration test
- [ ] Copyright header
- [ ] Code formatting
- [ ] Documentation
- [ ] Change log entry

### CODE REVIEW

File: akvo/rsr/models/user.py [hunk 1]
```diff
@@ -104,6 +104,18 @@ class User(AbstractBaseUser, PermissionsMixin):
...
 
+    def check_password(self, raw_password):
+        """
+        Returns a boolean of whether the raw_password was correct. Handles
+        hashing formats behind the scenes.
+        """
+        from akvo.rsr.models.login_log import is_password_deactivated
+        from django import forms
+        if is_password_deactivated(self.user.username):
+            raise forms.ValidationError('Password has been deactivated for XXX minutes')
```
Comment (zzgvh) [inline akvo/rsr/models/user.py]: FIXME: translation

File: akvo/rsr/tests/models/test_login_logging.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        # When
+        with self.assertRaises(ValidationError):
+            self.c.login(username=self.email, password='')
+
+        # Then
+        logs = LoginLog.objects.filter(email=self.email)
+        self.assertTrue(logs.exists())
+        self.assertFalse(logs.first().success)
+
+    def test_password_deactivates_after_max_attemtpts(self):
```
Comment (zzgvh) [inline akvo/rsr/tests/models/test_login_logging.py]: test_password_deactivates_after_max_attempts

File: akvo/rsr/models/login_log.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                'count': remaining_logins,
+                'time': PASSWORD_DEACTIVATION_TIME / 60.0
+            },
+            code='invalid_login',
+            params={'username': 'username'},
+        )
+
+    elif failed_logins >= MAX_FAILED_LOGINS:
+        raise forms.ValidationError(
+            _(u'Password has been deactivated for %(time)d minutes') % {
```
Comment (zzgvh) [inline akvo/rsr/models/login_log.py]: _(u'The account has been deactivated for %(time)d minutes') % {

File: akvo/rsr/tests/models/test_login_logging.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        # Given
+        for _ in range(MAX_FAILED_LOGINS + 10):
+            with self.assertRaises(ValidationError):
+                self.c.login(username=self.email, password='')
+        # HACK: Set the creation time of these login attempts to yesterday
+        creation_time = LoginLog.objects.first().created_at - timedelta(days=1)
+        LoginLog.objects.update(created_at=creation_time)
+
+        # When
+        self.c.login(username=self.email, password=self.password)
```
Comment (zzgvh) [inline akvo/rsr/tests/models/test_login_logging.py]: If I get this right, this assumes that LOGIN_DISABLE_TIME is no larger than 86400, but I think that's an OK assumption! 😜

Comment (punchagan) [reply akvo/rsr/tests/models/test_login_logging.py]: Nice catch. I changed this to remove the assumption, by going back by `LOGIN_DISABLE_TIME * 2` seconds

**Pull request merged**